packages/typescript/src/index.ts

```diff
@@ -42,11 +46,40 @@ export default function typescript(options: RollupTypescriptOptions = {}): Plugi
           parsedOptions,
           writeFile(fileName, data) {
             emittedFiles.set(fileName, data);
+          },
+          status(diagnostic) {
+            // Fullfil watcher promises by Typescript diagnostic message codes
+            if (diagnostic.category === ts.DiagnosticCategory.Message) {
+              switch(diagnostic.code) {
+                case 6032:
+                  tsWatcherStartResolve && tsWatcherStartResolve(false);
+                  tsWatcherStartPromise = null;
+                  tsWatcherStartResolve = null;
+                  break;
+
+                case 6194:
+                  tsWatcherFinishResolve && tsWatcherFinishResolve();
+                  tsWatcherFinishPromise = null;
+                  tsWatcherFinishResolve = null;
+                  break;
+              }
+            }
           }
         });
       }
     },
 
+    watchChange(id) {
+      if (!filter(id)) return;
+
+      // Race watcher start promise against a timeout in case Typescript and Rollup change detection is not in sync
+      tsWatcherStartPromise = Promise.race<Promise<void | boolean>>([
+        new Promise(resolve => setTimeout(resolve, 1000, true)),
+        new Promise(resolve => tsWatcherStartResolve = resolve)
+      ]);
+      tsWatcherFinishPromise = new Promise(resolve => tsWatcherFinishResolve = resolve);
+    },
+
     buildEnd() {
       if (process.env.ROLLUP_WATCH !== 'true') {
```

Comment:
Btw: This doesn't work if you start 'rollup.watch' process.env.ROLLUP_WATCH will be undefined

Reply:
https://github.com/rollup/rollup/pull/3616 will fix it.

Should be `if (this.meta.watchMode !== true)` then